plugins/modules/s3_object.py

```diff
@@ -636,86 +747,138 @@ def path_check(path):
+        "serversideencryption": "ServerSideEncryption",
+        "storageclass": "StorageClass",
+        "ssecustomeralgorithm": "SSECustomerAlgorithm",
+        "ssecustomerkey": "SSECustomerKey",
+        "ssecustomerkeymd5": "SSECustomerKeyMD5",
+        "ssekmskeyid": "SSEKMSKeyId",
+        "websiteredirectlocation": "WebsiteRedirectLocation",
+    }
 
     if temp_option in allowed_extra_args:
         return allowed_extra_args[temp_option]
 
 
-def upload_s3file(module, s3, bucket, obj, expiry, metadata, encrypt, headers, src=None, content=None, acl_disabled=False):
+def upload_s3file(
+    module,
+    s3,
+    bucket,
+    obj,
+    expiry,
+    metadata,
+    encrypt,
+    headers,
+    src=None,
+    content=None,
+    acl_disabled=False,
+):
     if module.check_mode:
-        module.exit_json(msg="PUT operation skipped - running in check mode", changed=True)
+        module.exit_json(
+            msg="PUT operation skipped - running in check mode", changed=True
+        )
     try:
-        extra = {}
-        if encrypt:
-            extra['ServerSideEncryption'] = module.params['encryption_mode']
-        if module.params['encryption_kms_key_id'] and module.params['encryption_mode'] == 'aws:kms':
-            extra['SSEKMSKeyId'] = module.params['encryption_kms_key_id']
-        if metadata:
-            extra['Metadata'] = {}
-
-            # determine object metadata and extra arguments
-            for option in metadata:
-                extra_args_option = option_in_extra_args(option)
-                if extra_args_option is not None:
-                    extra[extra_args_option] = metadata[option]
-                else:
-                    extra['Metadata'][option] = metadata[option]
-
-        if module.params.get('permission'):
-            permissions = module.params['permission']
+        extra = get_extra_params(
+            encrypt,
+            module.params.get("encryption_mode"),
+            module.params.get("encryption_kms_key_id"),
+            metadata,
+        )
+        if module.params.get("permission"):
+            permissions = module.params["permission"]
             if isinstance(permissions, str):
-                extra['ACL'] = permissions
+                extra["ACL"] = permissions
             elif isinstance(permissions, list):
-                extra['ACL'] = permissions[0]
-
-        if 'ContentType' not in extra:
-            content_type = None
-            if src is not None:
-                content_type = mimetypes.guess_type(src)[0]
-            if content_type is None:
-                # s3 default content type
-                content_type = 'binary/octet-stream'
-            extra['ContentType'] = content_type
-
-        if src is not None:
-            s3.upload_file(Filename=src, Bucket=bucket, Key=obj, ExtraArgs=extra)
+                extra["ACL"] = permissions[0]
+
+        extra["ContentType"] = get_content_type(
+            src, present=extra.get("ContentType")
+        )
```

Comment:
This changes the effect of the previous code and replaces the contents of extra["ContentType"] with the return value of get_content_type, which if extra already has "ContentType" (ie. present is truthy) returns None. 

I think we should remove the `present` argument from  `get_content_type` and then only call it if it's not present:

```python
if "ContentType"  not in extra:
  extra["ContentType"] = get_content_type(src)
```